### DISCUSSION TITLE
CT Module seeking

### DISCUSSION THREAD

**Discussion by crcrewso** [jrkerns/pylinac#364] [Ideas]
I'm wondering if there should be a revisit to the way the CatPhan and other phantom's modules are calculated. I am sure it would be a bit harder but what if 

- First we scan for the easiest to identify module
- search based on this guess for a region where the 'reference' module of the phantom would be (for CatPhan this is the resolution module) 
- Analyze a series of slices for the 'best' one, assign this slice as new reference
- Calculate from reference image in distance units (ideally mm)

**Comment by jrkerns** [comment]
This is pretty close to what pylinac does already:

> First we scan for the easiest to identify module

This is what happens. pylinac looks for the 404 module: https://github.com/jrkerns/pylinac/blob/master/pylinac/ct.py#L1028

> search based on this guess for a region where the 'reference' module of the phantom would be (for CatPhan this is the resolution  module)

I'm not sure how this is different than the first point?

> Analyze a series of slices for the 'best' one, assign this slice as new reference


What does the "best" one mean? If the best one is at the edge of a module, then it's possible all offsets will not be correct.

> Calculate from reference image in distance units (ideally mm)

This is what pylinac does. https://github.com/jrkerns/pylinac/blob/master/pylinac/ct.py#L1421-L1426

**Reply by crcrewso** [reply]
I'm sorry, but it's not quite what pylinac does, Based on my reading of your code Pylinac uses what it first finds as the center of 404 as the reference module. Let's call this the search once algorithm. 

I've written out the summary as a more general description. For the CatPhans, the locations would be recalculated based on the location of the slice that has the best result of CTP 528. So search twice would be search for 404, to use as a guess for 528. Search for 528. Recalculate all slices from 528. if we base everything off of slice width then we could pick the reference slices based on manufacturers reference ideology. 

I know this is a lot harder to maintain, but it would be more in keeping with the intent of the phantoms.

**Reply by jrkerns** [reply]
I didn't know there was a reference idealogy for the phantoms. Where can I find that info?

**Reply by crcrewso** [reply]
It was hard to find, I think I might be reading too many machinists manuals. For any of the fixed phantoms there's a defined reference that all other measurements are based on, this is important for confirming the compliance of the phantom geometrically when accepting it. 

Catphan 504 Page 7
https://static1.squarespace.com/static/5367b059e4b05a1adcd295c2/t/58b5c7001b631bbb2f82c584/1488307968931/Test+CTP+504+Manual+12_15.pdf

this catphan defines High res as the reference for all distance measurements in the phantom axis

Catphan 604 Page 8
https://static1.squarespace.com/static/5367b059e4b05a1adcd295c2/t/56a78807c21b869ca68747b1/1453819924918/CTP604+Manual+9-15.pdf

This catphan defines the HU slice module as the reference.

**Reply by jrkerns** [reply]
504 seems to use the 404: "The Catphan®504 phantom is designed so all test sections can be located by precisely 
indexing the table from the center of section CTP404 to the center of the other test 
modules" 404 is the HU module (page 9). 

604 also seems to use the same: "The Catphan®604 phantom is designed so all test sections can be located by precisely 
indexing the table from the center of scan section 2 to the center of the other test sections 
and low contrast module. " Section 2 from the image is the HU module. (page 9 & 10).

**Reply by crcrewso** [reply]
Looks like you are spot on and I'm misreading the manual. I'll chalk it up to the difference between diagram representation and actual intent. 

I'm still thinking the default slice detection could use some work (and I'd like to parallelize parts of this code) but I'll chalk it up as a nice to have. 

Also right now there's a lot of references to the modules by their CatPhan 504 names, even though the numbers change in some alternate phantoms. This is a bit of refactoring I might try to take on too.
